engines/ic3base.cpp

```diff
@@ -641,42 +645,35 @@ bool IC3Base::is_blocked(const ProofGoal * pg)
 
 bool IC3Base::propagate(size_t i)
 {
-  assert(i + 1 < frames_.size());
+  assert(!solver_context_);
+  assert(i < frontier_idx());
 
-  vector<IC3Formula> to_push;
   vector<IC3Formula> & Fi = frames_.at(i);
 
-  push_solver_context();
-  assert_frame_labels(i);
-  assert_trans_label();
-
   size_t k = 0;
+  IC3Formula gen;
   for (size_t j = 0; j < Fi.size(); ++j) {
-    const Term & t = Fi.at(j).term;
-
-    // Relative inductiveness check
-    // Check F[i] /\ t /\ T /\ -t'
-    // NOTE: asserting t is redundant because t \in F[i]
-    push_solver_context();
-    solver_->assert_formula(solver_->make_term(Not, ts_.next(t)));
-
-    Result r = check_sat();
-    assert(!r.is_unknown());
-    if (r.is_unsat()) {
-      to_push.push_back(Fi.at(j));
+    const IC3Formula & c = Fi.at(j);
+    assert(c.disjunction);
+    assert(c.term);
+    assert(c.children.size());
+
+    // NOTE: rel_ind_check works on conjunctions
+    //       need to negate
+    if (rel_ind_check(i + 1, ic3formula_negate(c), gen, false)) {
+      // can push to next frame
+      // got unsat-core based generalization
+      assert(gen.term);
+      assert(gen.children.size());
+      constrain_frame(i + 1, ic3formula_negate(gen), false);
```

Comment:
If I understand correctly, we are doing generalization now -- we were not doing it before. RIght?

Reply:
Yeah that's right. Not the expensive minimizing generalization, but just the cheap unsat-core based generalization that happens automatically in `rel_ind_check`